[Feature] Load BIP-85 Child Seeds

Chaitanya-Keyal opened the pull request:
## Description

Closes #354.

This PR introduces support for directly loading BIP-85 generated seeds into the UX.

- Adds `SeedBIP85FinalizeView`, with [this updated flow](https://excalidraw.com/#json=tbXTKuqraFUOCbBPMh9xj,LERaz9c9GSGufgOKdkPQCg) in mind.
- Adds `BIP85ChildSeed` and refactors codebase to eliminate the need of passing around `bip85_data`.
- Discards child seeds if parent is discarded.
- Adds visual indicators on all list views to differentiate child seeds and understand hierarchy.
	> Note: I initially wanted to use [this icon](https://fontawesome.com/icons/arrow-turn-down-right?f=classic&s=solid) instead of the plain arrow, but unfortunately it's a _pro_ icon. Perhaps we could make a custom icon for this in the SeedSigner Font?
- Adds child seed info in `SeedOptionsView`.
- Adds a new setting option for BIP-85.

### New Screenshots

<img width="240" height="240" alt="SettingsEntryUpdateSelectionView_bip85_child_seeds" src="https://github.com/user-attachments/assets/ffff70da-bf41-4d98-b460-52ecfb42d54c" /> <img width="240" height="240" alt="SeedBIP85FinalizeView_loadable" src="https://github.com/user-attachments/assets/acad4839-5d55-403f-8b56-824ad86e8d09" /> <img width="240" height="240" alt="SeedFinalizeView_bip85_child_seed" src="https://github.com/user-attachments/assets/69eb8224-a499-4a67-b179-7e0af5a82a21" /> <img width="240" height="240" alt="SeedOptionsView_bip85_child_seed" src="https://github.com/user-attachments/assets/05cc7f6c-6b2d-4a80-b565-826fb127a2c6" /> <img width="240" height="240" alt="SeedsMenuView_bip85_child_seed" src="https://github.com/user-attachments/assets/4434eacf-dbae-4753-b7f9-d4c3f4bcfefc" /> <img width="240" height="240" alt="SeedDiscardView" src="https://github.com/user-attachments/assets/ba1dc4fc-79ef-4dd0-bc6e-1e6cac04f80b" /> <img width="240" height="240" alt="SeedWordsView_bip85_child_seed" src="https://github.com/user-attachments/assets/07746c9f-49c9-4566-a8f9-dfacba06fc32" />


### Other minor changes:
- Rewords "Wipe" to "Discard" in `SeedDiscardView`.
- Makes the following `SelectSeedViews` consistent with other In-Memory Seeds views and use white for the fingerprint icon instead of blue.

<img width="240" height="240" alt="SeedDiscardView" src="https://github.com/user-attachments/assets/18afa8b5-8d59-4d2c-ac03-dc50d2d982af" /> <img width="240" height="240" alt="SeedSelectSeedView_address_verification" src="https://github.com/user-attachments/assets/bf974e21-abc9-4469-825d-166b32e11b0a" /> <img width="240" height="240" alt="SeedSelectSeedView_sign_message" src="https://github.com/user-attachments/assets/5ee7c5bc-2e31-4caf-a0af-26b881545d08" />


## TODO
- [x] Settings Menu: Add a third option to enable BIP-85 seed generation and loading.
	- Note: The BIP-85 setting changed from enabled/disabled to a three-option select (Disabled / View-only / Loadable), so users who previously enabled it will find it effectively disabled until they re-select a mode.
- [x] Visual Indicators: Add metadata (e.g., parent fingerprint, child index) in `SeedOptionsView` and differentiate child seeds/show parent relationship in Seeds List Views.
- [x] Tests, Screenshot Generator: Implement appropriate unit and integration tests, add screenshots.

---

This pull request is categorized as a:

- [x] New feature

## Checklist

- [x] I’ve run `pytest` and made sure all unit tests pass before submitting the PR

If you modified or added functionality/workflow, did you add new unit tests?

- [x] Yes

I have tested this PR on the following platforms/os:

- [x] Raspberry Pi OS [Manual Build](https://github.com/SeedSigner/seedsigner/blob/dev/docs/manual_installation.md)

Chaitanya-Keyal committed:
feat: load BIP85 child seeds

Chaitanya-Keyal committed:
feat: load BIP85 child seeds

kdmukai commented on src/seedsigner/models/settings_definition.py:
```diff
@@ -51,6 +51,13 @@ class SettingsConstants:
 
+    BIP85__ENABLED_WITH_LOADING = "L"
+    ALL_BIP85_OPTIONS = [
+        (BIP85__ENABLED_WITH_LOADING, _mft("Enabled (With Loading)")),
+        (OPTION__ENABLED, _mft("Enabled (Generation Only)")),
```
Maybe:
* "Enabled: view-only seeds"
* "Enabled: loadable seeds"

Naming things is always so hard.

I would prefer the more conservative option to be listed first.

kdmukai replied:
I think for max clarity, the existing `ENABLED` option should be updated / upgraded to its own dedicated `BI85__` constant. It's no longer sufficient to just say "ENABLED". It's now: enabled FOR WHAT?

kdmukai commented on src/seedsigner/views/seed_views.py:
```diff
@@ -375,6 +375,11 @@ def run(self):
             return Destination(SeedReviewPassphraseView)
         
+        elif self.seed.is_bip85_child_seed:
+            return Destination(SeedBIP85FinalizeView, view_args={
+                "bip85_data": self.seed.bip85_data,
```
Why not just have `SeedBIP85FinalizeView` accept a `seed_num` like other Views and let it retrieve the `seed.bip85_data` on its own? A light `seed.is_bip85_child_seed` sanity check in there would catch any later misuses.

It's not a big deal but we should avoid passing around data that's already easily accessible elsewhere. Also just feels a little clunky and in some cases fragile to pass data to Views this way (though I'm sure there are places in the code where this does happen).

kdmukai commented on src/seedsigner/models/seed.py:
```diff
@@ -22,7 +22,9 @@ class Seed:
-                 wordlist_language_code: str = SettingsConstants.WORDLIST_LANGUAGE__ENGLISH) -> None:
+                 wordlist_language_code: str = SettingsConstants.WORDLIST_LANGUAGE__ENGLISH,
+                 is_bip85_child_seed: bool = False,
+                 bip85_data: dict = None) -> None:
```
In order to keep the core `Seed` class as simple as possible, I think I'd prefer to see the bip85 particulars pulled out into a separate:
```python
class BIP85ChildSeed(Seed):
```

See the `ElectrumSeed` class.

We wouldn't need either of the properties then. The `is_bip85_child_seed` test would just be `isinstance(seed, BIP85ChildSeed)`. And we wouldn't need to protect `bip85 data` behind the property getter.

Could even override `get_bip85_child_mnemonic()` to throw an Exception (I'd define a new one like `BIP85GrandchildNotAllowedException`).

kdmukai replied:
And/or override `bip85_supported` in the child class.

kdmukai commented on src/seedsigner/gui/screens/seed_screens.py:
```diff
@@ -436,16 +436,54 @@ def __post_init__(self):
 
+    def __post_init__(self):
+        self.show_back_button = True
+        self.title = _("Finalize Seed")
```
Title shouldn't be "Finalize" if we're in generate-but-don't-load mode. Maybe "BIP-85 Child Seed".

And I think the options in that case should be:
* "View mnemonic"
* "Done" (return to parent's Seed Options View)

We have a few dead-end UX flows that force the user to use the BACK arrow, but that should be avoided if possible.

(note that buttons should use "title case" where only the first word is capitalized)

kdmukai commented on src/seedsigner/views/seed_views.py:
```diff
@@ -1160,6 +1172,61 @@ def run(self):
+            seed_num = self.controller.storage.finalize_pending_seed()
+            return Destination(SeedOptionsView, view_args={"seed_num": seed_num}, clear_history=True)
+        
+        if button_data[selected_menu_num] == self.PASSPHRASE:
+            return Destination(SeedAddPassphraseView)
```
The flow might feel a little disjoint, but I think I'd prefer:
* Load child
* View mnemonic

Having three button options starts to crowd the screen. And in that context, "BIP-39 passphrase" already feels a bit odd sitting above "View mnemonic".

So if we removed it, "Load child" would then just route to the existing Finalize Seed View where the bip39 passphrase option is offered. The user is already used to seeing this View so they know what it's doing.

kdmukai commented:
In the current draft, a child seed can generate its own bip85 child seeds. That is absolutely unacceptable to me and will need a check to prevent it.

kdmukai commented:
Also definitely need the TODO bullet point you list in the description to differentiate the child seeds in the list of seeds loaded in memory.

Chaitanya-Keyal committed:
feat(BIP85): add visual indicators for child seeds

Chaitanya-Keyal committed:
feat(BIP85): add visual indicators for child seeds

Chaitanya-Keyal committed:
test: add BIP85 load child flow test

Chaitanya-Keyal committed:
test: add BIP85 load child flow test

Chaitanya-Keyal committed:
test(screenshot-generator): add screenshots for new BIP85 screens

Chaitanya-Keyal committed:
test(screenshot-generator): add screenshots for new BIP85 screens

Chaitanya-Keyal commented:
Thanks for the review, Keith. I’ve completed all my remaining TODO items for this PR and incorporated all the changes you suggested. The PR description has been updated to reflect the new changes, and this is now ready for another review.

A couple of points for discussion:
- What do we think about the “button” I’ve used in the `SeedOptionsView` to display the child seed’s information?
- Is the current arrow icon (in seed list views) sufficient, or should we switch to the alternative icon I mentioned in the PR description? If we decide to switch, could I get some guidelines on how to add an icon to the SeedSigner font?

kdmukai commented:
> * What do we think about the “button” I’ve used in the `SeedOptionsView` to display the child seed’s information?

It being info only and not actually navigating anywhere is definitely a violation of our UX rules.

I think the parentage is well represented Seeds listing. So I think the `SeedOptionsView` title can just focus on what's unique to the child:
* "Child 14: abcd1234" (child's fingerprint)
* Or maybe just "14: abcdd1234" (child's fingerprint)

Then that info-only button isn't necessary anymore.

kdmukai commented:
`SeedSelectSeedView` appends seeds and children as they're loaded (note: children DO correctly stay with their parent even when multiple seeds are loaded and children of different seeds).

But I wonder if the children should be sorted by their child index instead.

```
Seed A
 -> #13: abcd1234
 -> #5: abcd1234
 -> #423: abcd1234
Seed B
 -> #11: abcd1234
 -> #2: abcd1234
```

vs

```
Seed A
 -> #5: abcd1234
 -> #13: abcd1234
 -> #423: abcd1234
Seed B
 -> #2: abcd1234
 -> #11: abcd1234
```

kdmukai commented:
I don't think altering the fingerprint color for child seeds is the best way to go. If there is any differentiation, maybe it should be via a custom icon designed by @easyuxd (see the "Finalize Child Seed" screenshot in the PR description). At the very least I'd leave it to Easy to weigh in on the color (stick with fingerprint blue or a diff alt color?).

The arrows in "Select Signer" and "Verify Address" should also be reviewed. I think the arrow icon itself is fine, though something more like ↳ shaped might be better. And I don't think it should be shaded blue in "Verify Address". In fact, it's probably an implementation miss that one screen renders the fingerprints in the buttons white while in the other it's blue.

Another thought: maybe the SeedOptionsView for a child (the "c1969012" screenshot) could replace the fingerprint icon with the big "#" icon and then add the child index number as I suggested above to give us: "#14: c1969012"

kdmukai commented on src/seedsigner/controller.py:
```diff
@@ -8,7 +8,7 @@
 from seedsigner.gui.toast import BaseToastOverlayManagerThread
 from seedsigner.models.psbt_parser import PSBTParser
-from seedsigner.models.seed import Seed
+from seedsigner.models.seed import BIP85ChildSeed, Seed
```
Looks like this isn't necessary?

kdmukai commented on src/seedsigner/gui/components.py:
```diff
@@ -190,6 +190,8 @@ class FontAwesomeIconConstants:
     SQUARE_CARET_UP = "\uf151"
     UNLOCK = "\uf09c"
     X = "\u0058"
+    INDEX = "\u0023"
```
Font Awesome refers to this as "hashtag". I usually refer to it as "pound", though I think that might be a little outdated. I'd probably go with `HASHTAG` here.

kdmukai commented on src/seedsigner/gui/screens/seed_screens.py:
```diff
@@ -438,14 +444,55 @@ def __post_init__(self):
+            label_text=_("child fingerprint"),
+            value_text=self.child_fingerprint,
+            font_size=GUIConstants.get_body_font_size(),
+            screen_x=GUIConstants.LIST_ITEM_PADDING,
```
I wouldn't use this constant here. It's meant to specify vertical spacing between list items. Even if it's the exact right pixel value needed here, it's still best to avoid using layout constants in ways that weren't intended (though, ya, I'm sure there are other misuses in the existing code). If that layout constant is ever changed, we do our best to guarantee that it won't have other unintended effects.

kdmukai commented on src/seedsigner/gui/screens/seed_screens.py:
```diff
@@ -438,14 +444,55 @@ def __post_init__(self):
+        self.child_fingerprint_icontl = IconTextLine(
+            icon_name=SeedSignerIconConstants.FINGERPRINT,
+            icon_color=GUIConstants.INFO_COLOR,
+            icon_size=GUIConstants.ICON_FONT_SIZE + 12,
```
The existing call on line 436 ideally wouldn't have hard-coded a pixel size adjustment value. It's best to avoid replicating it here since we'll eventually need to refactor all the hard-coded pixel values. In general, a relative value (e.g. 1.5x the size of the accompanying font) will be the most flexible.

Chaitanya-Keyal commented:
Thanks again, Keith. I’ve [made all the suggested changes](https://github.com/SeedSigner/seedsigner/compare/48be674658d57c7a9e4a8cdc0b53ecd207ed0341~4...48be674658d57c7a9e4a8cdc0b53ecd207ed0341) (also updated screenshots in the PR desc):
- Child seeds are now sorted by index
- The info-only button has been removed, and the `SeedOptionsView` title now includes the child index with the `#` icon

On the icons:
- I agree that the [alternative arrow icon](https://fontawesome.com/icons/arrow-turn-down-right?f=classic&s=solid) mentioned in the PR desc would be a better fit, though it would also need to be custom
- A custom icon for child seeds would be the ideal solution (for both `SeedFinalizeView` and `SeedOptionsView`). The current `#` works as a placeholder, but something more tailored would definitely look better

In the meantime, until Easy has time to design the new icons, I’m also open to reverting to the fingerprint icon for the `SeedOptionsView` and adjusting its color if needed.

kdmukai commented on src/seedsigner/gui/screens/seed_screens.py:
```diff
@@ -416,18 +416,24 @@ def _run(self):
-            icon_color=GUIConstants.INFO_COLOR,
-            icon_size=GUIConstants.ICON_FONT_SIZE + 12,
+            icon_color=icon_color,
+            icon_size=GUIConstants.ICON_FONT_SIZE * 1.5,
```
I'm not sure how PIL handles non-integer sizes or if it matters at all. But we always cast to `int` just to be safe.

Chaitanya-Keyal replied:
Missed this, my bad.

kdmukai commented on src/seedsigner/models/seed.py:
```diff
@@ -239,3 +239,26 @@ def seedqr_supported(self) -> bool:
+
+    @property
+    def bip85_supported(self) -> bool:
+        return False
```
A short explanatory comment might help here.

kdmukai commented on src/seedsigner/models/settings_definition.py:
```diff
@@ -51,6 +51,14 @@ class SettingsConstants:
+    BIP85__ENABLED_VIEW_ONLY = "V"
+    BIP85__ENABLED_LOADABLE = "L"
+    ALL_BIP85_OPTIONS = [
+        (BIP85__ENABLED_VIEW_ONLY, _mft("Enabled: view-only seeds")),
+        (BIP85__ENABLED_LOADABLE, _mft("Enabled: loadable seeds")),
```
I feel like "seeds" is a little less clear than something like:
* "child seeds" (probably the most clear)
* Or even just "children"

Chaitanya-Keyal replied:
<img width="240" height="240" alt="image" src="https://github.com/user-attachments/assets/a0d51b8e-f3b4-4434-886a-af90a9871264" />

The settings title does mention BIP-85 **child** seeds, would this not feel a bit repetitive?

kdmukai replied:
> The settings title does mention BIP-85 **child** seeds, would this not feel a bit repetitive?

That's a good point. Though at the same time I'm always happy to lean towards being as explicit as possible, even if it gets a little redundant (e.g. "passphrase" is such a messy term that I always insist on saying "BIP-39 passphrase" over and over).

I think we can drop "Enabled" here; that is implicit, especially when "Disabled" is an option.

Also: The Silent Payments PR lists "Disabled" first since that's the default. And in this case, "Disabled" isn't just the default, in my mind it's strongly recommended. So I'd consider listing "Disabled" first to drive home that point. It's a tiny break from our usual Enabled/Disabled convention, but this list isn't strictly Enabled/Disabled so I think there's some wiggle room here.

If it doesn't look strange without "Enabled", then with the extra room, I think I'd still prefer the more explicit:
* "View-only child seeds"
* "Loadable child seeds"

But just a mild opinion. This is not a make-or-break to me.

Chaitanya-Keyal replied:
Dropping "Enabled" sounds looks a good choice. Including "child" seeds then doesn't feel like a really long option. I've updated the options and reordered them:
<img width="240" height="240" alt="SettingsEntryUpdateSelectionView_bip85_child_seeds" src="https://github.com/user-attachments/assets/fd1b3c79-6ee9-4c1f-820b-e34f023676c8" />

kdmukai commented on src/seedsigner/views/psbt_views.py:
```diff
@@ -32,13 +33,20 @@ def run(self):
+            if isinstance(seed, BIP85ChildSeed):
+                # Indent the Child Seed Label for a better hierarchical view
+                icon = " " * (GUIConstants.EDGE_PADDING // 4) + FontAwesomeIconConstants.ARROW_RIGHT
+                label = f"#{seed.child_index}: " + fingerprint
```
This probably needs to be wrapped for translation via `_()`. It's possible that other languages might not use the "#" sign (ya, but we are already using the HASHTAG icon... 🤷‍♂️) and maybe the colon could be an issue, too?

kdmukai commented on src/seedsigner/views/psbt_views.py:
```diff
@@ -32,13 +33,20 @@ def run(self):
                 # Doesn't look like this seed can sign the current PSBT
                 # TRANSLATOR_NOTE: Inserts fingerprint w/"?" to indicate that this seed can't sign the current PSBT
-                button_str = _("{} (?)").format(button_str)
+                label = _("{} (?)").format(label)
```
Oof, but not sure if there are any concerns about passing this through another round of translation lookups...

I vaguely remember that you could have a rare case where:

Lang1 "A" -> Lang2 "B"
But "B" just happens to have its own Lang1 translation to "C".

English "bread" -> Spanish "pan"
English "pan" (e.g. frying pan) -> Spanish "sartén"

I don't think there's any risk of that inadvertent "C" outcome here. Just a little head-spinny to think through.

Chaitanya-Keyal replied:
Damn, hadn't thought of this before. But yeah you're right, in this case it's the index and fingerprint, so there shouldn't be any issues.

kdmukai commented on src/seedsigner/views/tools_views.py:
```diff
@@ -478,8 +478,15 @@ def run(self):
+            fingerprint = seed.get_fingerprint(self.settings.get_value(SettingsConstants.SETTING__NETWORK))
+            if isinstance(seed, BIP85ChildSeed):
+                # Indent the Child Seed Label for a better hierarchical view
+                icon = " " * (GUIConstants.EDGE_PADDING // 4) + FontAwesomeIconConstants.ARROW_RIGHT
```
I wonder if PIL handles "\t". I think I remember trying it but it just treats it as a single whitespace.

Chaitanya-Keyal replied:
<img width="175" height="60" alt="image" src="https://github.com/user-attachments/assets/891c5752-6ae7-4dac-9141-f04e6f893973" />

`"\t"` was my first instinct here too, but this is what we get when using `"\t" + FontAwesomeIconConstants.ARROW_RIGHT`. I was a bit surprised by this tbh.

kdmukai commented:
This will be a bit of annoying fussing about, but I'd prefer that the various screenshots that show the in-memory seeds list (there are about 4-5) stay "clean" with no bip85 children in the list. The rationale is that we'll be constantly leveraging these screenshots for tutorials and other materials so we should always have a beginner-friendly "clean" screenshot. And then add in whatever more advanced feature screenshots / variants we might need (e.g. PSBTOverviewView_op_return).

So I would just create a single alternate version of "In-Memory Seeds" that includes a child. I'd place it just after SeedOptionsView_bip85_child_seed.

I don't think we need a bip85'ed variant for all the other seed list screenshots.

kdmukai commented on tests/test_flows_seed.py:
```diff
@@ -477,6 +477,32 @@ def load_right_seed_into_decoder(view: View):
+                FlowStep(seed_views.SeedBIP85SelectChildIndexView, screen_return_value=0),
+                FlowStep(seed_views.SeedBIP85FinalizeView, button_data_selection=seed_views.SeedBIP85FinalizeView.LOAD),
+                FlowStep(seed_views.SeedFinalizeView, button_data_selection=seed_views.SeedFinalizeView.FINALIZE),
+                FlowStep(seed_views.SeedOptionsView),
```
I would either extend this flow test (easier option) or write a separate one (additional harness work) to also test the bip85 parent delete flow to ensure that the child seeds are properly removed.

kdmukai commented on src/seedsigner/views/seed_views.py:
```diff
@@ -1271,12 +1356,8 @@ def __init__(self, seed_num: int, bip85_data: dict = None, confirmed_list: list[
-            self.mnemonic_list = self.seed.get_bip85_child_mnemonic(self.bip85_data["child_index"], self.bip85_data["num_words"]).split()
-        else:
-            self.mnemonic_list = self.seed.mnemonic_display_list
+        self.mnemonic_list = self.seed.mnemonic_display_list
```
Ooh, this and the related edits are a nice simplification!

kdmukai commented on src/seedsigner/views/seed_views.py:
```diff
@@ -495,6 +526,10 @@ def run(self):
             else:
                 self.controller.storage.clear_pending_seed()
+
+            for child_seed in self.bip85_child_seeds:
+                self.controller.discard_seed(self.controller.storage.seeds.index(child_seed))
```
It's pretty minor, but this feels like something that should be done in `Controller.discard_seed()`. Not sure if it should then pass that on to a call in `SeedStorage`, but either way I don't think this level of internal management should be done in the View.

Pulling this into `Controller.discard_seed()` would have the added benefit of being something that could be unit tested then, too. And would simplify much of this View.

Chaitanya-Keyal replied:
I had originally implemented it this way, but later changed my mind for some reason. That’s probably what caused the unused import in `controller.py` you mentioned earlier 😅.

Chaitanya-Keyal committed:
test: add tests for `SeedStorage`

Chaitanya-Keyal committed:
test: add tests for `SeedStorage`

Chaitanya-Keyal commented:
Here’s the [latest set of changes](https://github.com/SeedSigner/seedsigner/compare/08fc4ef879c78c6adaa6f1a62ed29cea74cc53e2~10...08fc4ef879c78c6adaa6f1a62ed29cea74cc53e2) after addressing the review comments:

- Refactored all seed list views to have a more DRY approach and use the new `Seed` class attributes for label and icon
- Updated the Screenshot Generator to have only one screenshot with a BIP85 child seed loaded.
- Moved the discard seed logic into `SeedStorage`
- Added tests for both `discard_seed` and for verifying the order of seeds set by `finalize_pending_seed`
- Child Seeds are now also sorted by number of words if their index is the same

Also, I've rebased the branch onto `dev` and cleaned up the commit history.

Chaitanya-Keyal commented on src/seedsigner/views/seed_views.py:
```diff
@@ -1174,7 +1179,6 @@ def __init__(self, seed_num: int, num_words: int):
 
 
     def run(self):
-        # TODO: Change this later to use the generic Screen input keyboard
```
Note: The `TODO` comment was removed since the generic `KeyboardScreen` is already in use. Git history shows it was introduced in #194 and became obsolete after the refactor in #196, but the comment was mistakenly left behind.

Chaitanya-Keyal committed:
fix(BIP85): use `Seed` object instead of `seed_num` for storing parent relation

Chaitanya-Keyal commented:
#810 is now merged so the following comment is obsolete, the seed object is used everywhere now instead of seed num.

---

I realized there's an issue with storing `parent_seed_num` in `BIP85ChildSeed`.

Say the following seeds are loaded:

```
0. Seed A
1. Seed B
2.  -> #0: abcd1234
```

In this case, the `parent_seed_num` for `abcd1234` is `1`.

Now, suppose the user loads a child seed for Seed A:

```
0. Seed A
1. -> #0: 1234abcd
2. Seed B
3. -> #0: abcd1234
```

Here, the `parent_seed_num` for `abcd1234` still says `1`, even though it has actually changed to `2`. An even simpler example is if Seed A were discarded.

The point is that we should **not** store `seed_num` as an identifying value for a seed.

**I’ve updated the PR to store the parent `Seed` object itself, so this issue should not occur anymore.**


That said, I also think we should avoid passing around `seed_num` elsewhere as well. Technically, `seed_num` is only used _temporarily_ across views. Its value is based on the index of the button list from which the seed was selected, and any flow after that cannot modify the seeds list (except for this new BIP85 flow and the simple "add new seed" flow). In the "add new seed" case, the seed is simply appended, so the old `seed_num` references remain valid; and the flow ends afterward, effectively resetting any temporary `seed_num` usage.

In my opinion, it would be much more robust to use the actual `Seed` object as a reference instead of its index in the list. ~I'll try refactoring the codebase and open a separate PR for this.~ I've opened PR #810 that refactors the codebase to use this approach.

Looking ahead, if we ever support newer seed types, having direct `Seed` object references also seems like a win(?). And if for any reason the seeds list ever stops being strictly ordered (say we decide to use a different data structure), using indices would break things.

SatMeNow commented:
@Chaitanya-Keyal I just reviewed this and tested it on my device. Looks quite great 👍
I know that there are arguments against BIP-85 because of it's complexity. The _view-only child seed_ is a good compromise in my optinion 👍
Hope to see this in the next release.

One thing on _child seeds of child seeds_:
I can live without it! I even would not use it. But I prefer not to limit any users here! Users should know what they do. And there will be security setups where users will benefit from _unleashed_ BIP-85.

Chaitanya-Keyal commented:
Rebased to current `dev` and updated the screenshot generator changes to follow the new "Mock" pattern introduced in #860

bjrb20 commented:
I built and tested this on my seed singer and it works great! BIP-85 is so useful and this makes it really easy to managed multiple seeds 👍

newtonick commented on src/seedsigner/models/seed.py:
I believe it would be best to update this equal operator to treat `Seed` and `BIP85ChildSeed` as not equal.

The reason, even though an edge case, is they can contain identical seed bytes, but they represent different in-memory entries. A BIP85ChildSeed carries parent/index provenance and participates in parent-child display and discard behavior. With the current value-only equality, this sequence creates inconsistent state. For example:

1. Load a child mnemonic independently as a regular Seed.
2. Then the same mnemonic/seed is later derived and loaded as a BIP85ChildSeed.
3. SeedStorage.finalize_pending_seed() sees the two objects as equal and does not insert the child object.
4. But it does returns the non-stored BIP85ChildSeed, so the immediate UI presents it as a child even though storage still contains only the regular seed. Which is strange. It temp a BIP85 Child but as soon as you back out, it's gone.

By treating the two representations as unequal would allow both entries to appear in the list at the same time. One as the independently loaded seed and one under its BIP-85 parent. Discarding the parent would then remove only the child entry while preserving the independently loaded seed. It would also make storage lookups and removal unambiguous when both representations control the same keys.

The downside is the same seed would be in the list twice, however it would only happen if the user added the same seed twice explicitly. I prefer this over the inconsistent state case.

One possible implementation:

```python3
def __eq__(self, other):
     if isinstance(other, Seed):
         if isinstance(self, BIP85ChildSeed) != isinstance(other, BIP85ChildSeed):
             return False
         return self.seed_bytes == other.seed_bytes
     return False
```

Chaitanya-Keyal replied:
Agreed, and done. Implemented essentially as you wrote it. Also added a bunch of tests.

newtonick commented on src/seedsigner/models/seed.py:
```diff
@@ -239,3 +260,45 @@ def seedqr_supported(self) -> bool:
+        # TRANSLATOR_NOTE: Inserts BIP-85 child index and its fingerprint, e.g. "#3: abcd1234"
+        label = _("#{}: {}").format(self.child_index, fingerprint)
+
+        return label
```
I'm not a fan of having GUI code in the model layer. It breaks the model/GUI separation layer/architecture.

I don't have a great solution, but one thought is to create new screen to replace ButtonListScreen in psbt_view, seed_view, etc.

I shared this idea with AI and it suggested this partial implementation:

adding a seed specific button option in `seed_screens.py:`
```python3
@dataclass
class SeedButtonOption(ButtonOptionWithoutTranslation):
    child_index: int = None
```

ie:
- child_index is None → regular seed;
- child_index is not None → BIP-85 child.

Add `SeedButtonListScreen`:
```python3
@dataclass
class SeedButtonListScreen(ButtonListScreen):
    def __post_init__(self):
        for button_option in self.button_data:
            if not isinstance(button_option, SeedButtonOption):
                # Leave ordinary action buttons unchanged.
                continue

            if button_option.child_index is not None:
                # TRANSLATOR_NOTE: Inserts BIP-85 child index and fingerprint,
                # e.g. "#3: abcd1234"
                button_option.button_label = _("#{}: {}").format(
                    button_option.child_index,
                    button_option.button_label,
                )
                button_option.icon_name = (
                    " " * (GUIConstants.EDGE_PADDING // 4)
                    + FontAwesomeIconConstants.ARROW_RIGHT
                )
            else:
                button_option.icon_name = (
                    SeedSignerIconConstants.FINGERPRINT
                )

        super().__post_init__()
```


Example view layer:

```python3
button_data = []

for seed in seeds:
    child_index = None
    if isinstance(seed, BIP85ChildSeed):
        child_index = seed.child_index

    button_data.append(
        seed_screens.SeedButtonOption(
            button_label=seed.get_fingerprint(network),
            child_index=child_index,
        )
    )

...

selected_menu_num = self.run_screen(
    seed_screens.SeedButtonListScreen,
    title=_("In-Memory Seeds"),
    button_data=button_data,
    is_button_text_centered=False,
)
```

`PSBTSelectSeedView` would have some complications around handling the "?" to indicate if appears capable of signing the PSBT.

I believe all of these would need to be updated:
- SeedsMenuView → SeedButtonListScreen
- PSBTSelectSeedView → SeedButtonListScreen
- ToolsAddressExplorerSelectSourceView → SeedButtonListScreen
- SeedSelectSeedScreen → inherit from SeedButtonListScreen

... There would be more. I didn't finish the implementation. This approach is better because it keeps icons, translation, and indentation in the Screen layer. It's also DRY.

Chaitanya-Keyal replied:
Makes sense. I've updated the PR improving my implementation with your inputs. But I skipped creating the separate `SeedButtonListScreen` and instead only used `SeedButtonOption` with it decorating itself, so there's lesser changes required and still keeps everything DRY.

newtonick commented on src/seedsigner/views/seed_views.py:
```diff
@@ -1038,29 +1042,28 @@ class SeedWordsView(View):
+        if isinstance(self.seed, BIP85ChildSeed):
             # TRANSLATOR_NOTE: Inserts the child index (e.g. "Child #0")
-            title = _("Child #{}").format(self.bip85_data["child_index"])
+            title = _("Child #{}").format(self.seed.child_index)
```
When testing, I never saw this title for a Child seed. It always was showing me "Seed Words" instead of "Child #0". I didn't nail down exactly why. Might be an old bug.

Chaitanya-Keyal replied:
Good catch, and yes it's an old bug, not something this PR introduced. `SeedWordsScreen.__post_init__` unconditionally overwrites whatever title the View passes, so "Child #{}" never rendered before as well.

I've fixed it in this PR and `SeedWordsScreen` now takes a `bip85_child_index` and owns the string creation. Also added a `SeedWordsView_bip85_child_seed` screenshot so the title is actually covered going forward.
